### ISSUE TITLE
--auto-detect arg doesn't works as good as the standalone version

### ISSUE THREAD

**Issue by schmurtzm** [diablodiab/libretro-scummvm-backend#23]
Hi, this issue follow this ScummVM forum post : 
https://forums.scummvm.org/viewtopic.php?p=97569#p97569

**To resume :** 
When running  'Eye of the Beholder' without .scummvm file it doesn't work with core (`ERROR : failed to open EXE 'intro.EXE' !`) but it works with standalone.

By the way it could be useful to trigger `--auto-detect` when the .scummvm file is empty. And also we should be able to specify only a folder path as parameter to the core command line.

-----------------------------------------------

**Long version :** 
The scummVM dev team recommand to use the arg "`--auto-detect`" command line : it avoid to search for the shortname of the game to create the command line.

[This is how the command line is generated in the libretro core](https://github.com/diablodiab/libretro-scummvm-backend/blob/17c60f535fb6066e54ad8a8c5de82f00798edce4/libretro.cpp#L306) 
So if a .scummvm file exists then it run this command line :  `-p gamedir shortname`
and without .scummvm file it runs : `-p gamedir --auto-detect`

So when you run your scummvm core with a file which not exist like `./Roms/monkey2/monkey2.` in reality it runs scummVM with "--auto-detect" arg and the folder as parameter.

So you would say : "we should avoid to use .scummVM to use always the "`--auto-detect`" thing. ... yes but after some tests it works perfectly on standalone but with the core sometime it fails on some games.

**Example :**
_Indiana Jones and the Fate of Atlantis (this one has doublons) :_
- works fine without .scummvm file
- doesn't work with .scummvm file with inside atlantis (due to the doublon, scummVM doesn't know which engine to choose, devs plan to change "Unrecognized game XXX" to "Unrecognized or ambiguous game XXX" but it still visible only in logs concerning RA libretro core)
- works fine with .scummvm file with inside : scumm:atlantis

_Eye of the Beholder (no doublons) :_
- doesn't work without .scummvm file (works fine on windows standalone so it should work : `ERROR : failed to open EXE 'intro.EXE' !`)
- works fine with .scummvm file with eob
- works fine with .scummvm file with kyra:eob

**So to conclude** : The autodetect version could help a lot of users who are not aware of the .scummvm file shortcut creation. That's a shame that this autodetect feature doesn't work as well then standalone on the core version of ScummVM.

If we recommend to users to always create a .scummvm file with the long name (engine:shortname) then they will have the best compatibility. 

The problem is that the engine is not indicated in the [compatbility chart](https://www.scummvm.org/compatibility/DEV/) so it requires to run --list-games but when you use libretro core it's not very handy :P
EDIT : the [compatbility chart](https://www.scummvm.org/compatibility/DEV/)  now contains the required information with full name : useful for standalone and RA core ! Thanks to the team for that change :)

**Comment by ner00** [comment]
I'me confused about what the actual problem is. Could you give a practical example of an RA playlist entry, its `scummvm.ini` counterpart, and the game's `.scummvm` file and gameid? Also, none of my IDs have doublons, because all of them only use the game ID and never the engine ID nor a concatenation of both. The engine ID is only referenced inside `scummvm.ini`

I do agree that the playlist building is not ideal, it's manual and maybe that's your point. The process I follow is as laid in the libretro documentation:

1. Launch ScummVM core;
2. Add game to the list (scummvm.ini) from the GUI;
3. Create a `.scummvm` file inside the game folder using the game ID as its filename and also place the same game ID inside that file.
4. Scan the game folder with RetroArch to add the game entry to the playlist;
5. Map the game path to the `.scummvm` file manually in case the scan failed to do so automatically.

Now, if the scan process does not add the `.scummvm` file as the path file, there will be an issue as you mentioned, most definitely, and in that case one should edit the playlist and manually correct the entry to point to the game's `.scummvm` file. This can be an issue if you're adding games in bulk, and if that's your whole point then yes, there is much to improve there, and in general on the whole hybrid playlist method.

I'm genuinely asking because I have no issue as is, manual labor required notwithstanding. Here's the example of IJ: FoA:

**scummvm.ini**
```
[atlantis]
description=Indiana Jones and the Fate of Atlantis (CD/DOS/English)
extra=CD
speech_volume=192
path=F:\RetroArch\roms\ScummVM\Indiana Jones and the Fate of Atlantis (CD DOS)\
engineid=scumm
subtitles=true
gameid=atlantis
original_gui=true
language=en
enable_enhancements=true
mute=false
sfx_volume=192
music_volume=192
platform=pc
music_driver=auto
opl_driver=auto
speech_mute=false
guioptions=gameOption2 gameOption4 lang_English
```
**ScummVM.lpl**
```
    {
      "path": "F:\\RetroArch\\roms\\ScummVM\\Indiana Jones and the Fate of Atlantis (CD DOS)\\atlantis.scummvm",
      "label": "Indiana Jones and the Fate of Atlantis (CD/DOS)",
      "core_path": "cores\\scummvm_libretro.dll",
      "core_name": "ScummVM",
      "crc32": "60E9988F|crc",
      "db_name": "ScummVM.lpl"
    }
```
**F:\RetroArch\roms\ScummVM\Indiana Jones and the Fate of Atlantis (CD DOS)\atlantis.scummvm**
```
atlantis
```


I saw you mentioning elsewhere the case of Grim Fandango, which repeatedly checks the files because apparently it can't get or set the `check_gamedata` setting, which for me isn't the case at all and works perfectly fine. So, here's my example:

**scummvm.ini**
```
[grim-win]
description=Grim Fandango (Windows/English)
mute=false
rendering_mode=0
speech_volume=192
check_gamedata=false
directors_commentary=0
game_devel_mode=false
grim_language=0
datausr_load=false
talkspeed=153
path=F:\RetroArch\roms\ScummVM\Grim Fandango (Windows CD)
engineid=grim
subtitles=true
directors_commentary_volume=0
resolution_scaling=0
gameid=grim
advanced_lighting=0
language=en
widescreen=0
sfx_volume=192
savepath=F:\RetroArch\saves
music_volume=192
show_fps=false
platform=windows
music_driver=auto
opl_driver=auto
mouse_speed=0
grim_fullscreen=0
joystick=false
speech_mute=false
gamma=0
guioptions=sndNoMIDI gameOption1 gameOption2 lang_English
```

**ScummVM.lpl**
```
    {
      "path": "F:\\RetroArch\\roms\\ScummVM\\Grim Fandango (Windows CD)\\grim-win.scummvm",
      "label": "Grim Fandango (Windows/CD)",
      "core_path": "cores\\scummvm_libretro.dll",
      "core_name": "ScummVM",
      "crc32": "DETECT",
      "db_name": "ScummVM.lpl"
    }
```

**F:\RetroArch\roms\ScummVM\Grim Fandango (Windows CD)\grim-win.scummvm**
```
grim-win
```

**Comment by i30817** [comment]
You can use the manual scanner to scan for '.scummvm' extensions instead of using the normal and quite horrible RA scanner and you won't need to edit playlists. 

However i recommend you use my tool instead @ner00 : https://github.com/i30817/libretro-mkscumm installable in linux with `pip install libretro-scummvm-playlist`. It makes the playlists and scummvm files for you, and you 'only' need to scan/add the games inside the core GUI then delete the games which you don't want. This is still kind of annoying because scummvm mass add is flawed and sometimes adds games which you either don't want, or are actually wrong (the normal 'add' is well behaved but you'd go insane trying to add games one by one). For instance, Ultima 8 adds the 'actual game' and the 'USECODE' directory inside the game in mass add.

It's a bit annoying. But using my tool saves you time and also makes you have the 'correct' scummvm games names in RA playlists (with accents and 'forbidden windows characters'). You can also do that manually of course. But it's too much work.

**Comment by schmurtzm** [comment]
Hi, thanks for your answer. Firstly I don't use playlist at all ;)

To resume and keep it simple :  the `--auto-detect` feature is used by the core when you don't fill the .scummvm file. It's probably not recommended but still an interesting feature because it's simple to use.

In the case of Indiana Jones and the Fate of Atlantis, with an empty scummvm.ini file  `--auto-detect` works on standalone and don't run correctly in the core (it take the wrong engine).

**Comment by i30817** [comment]
The branch to use --auto-detect is not the .scummvm file being empty (note, i'm not mentioning the scummvm.ini, which may or may not be empty), it's the path being loaded _not_ being a .scummvm file (more specifically, it's the '.scummvm' string not being in the path, which is error prone but whatever):

https://github.com/diablodiab/libretro-scummvm-backend/blob/main/libretro.cpp#L306

And with a 'empty' scummvm.ini (not really empty, just without any game entries, it's a very very bad idea to start the core without at least some set properties because the default in scummvm is to have autosaves turned on, and the core _will_ crash if it can't find the directory to autosave), and the game fate of atlantis:

`retroarch -L scummvm .` inside the game directory, will run fate of atlantis (i just built the core).

I still think it's a terrible idea to use --auto-detect. And you _need_ a minimal scummvm.ini anyway, at least to prevent that save crash (among setting up other things like the soundfonts and mt32 rom).

**Comment by schmurtzm** [comment]
> The branch to use --auto-detect is not the .scummvm file being empty (note, i'm not mentioning the scummvm.ini, which may or may not be empty), it's the path being loaded _not_ being a .scummvm file (more specifically, it's the '.scummvm' string not being in the path, which is error prone but whatever):

Sorry I was talking about this case : when no .scummvm is given : 
https://github.com/diablodiab/libretro-scummvm-backend/blob/main/libretro.cpp#L333

**Comment by i30817** [comment]
Yes, that branch was what i mentioned is activated with `retroarch -L scummvm DIRECTORY` (i used . because i already was in the game directory for atlantis).

This runs the game, what's your problem? I have the cd version of fate of atlantis here, so maybe it's different for the floppy.

**Comment by schmurtzm** [comment]
Yes in fact it is an old issue from me, I was discovering ScummVM. Since I have understood that the right way is to use scummvm.ini as i30817 mention it.
We could close this issue, except the fact that the diffreence behavior of  `--auto-detect` between the standalone and the core is still strange. It should be indicated as not recommended in comment :p

**Comment by i30817** [comment]
BTW, a strange thing about using this way.... it actually deletes the game entry from scummvm.ini

So that's another reason not to use it. I think this is a upstream problem.

**Comment by schmurtzm** [comment]
That's true. Same behavior if you run the game with its "ScummVM ID" and not the "Target" (= the "scummvm.ini section name" that appear between []) 
Extract of a conversation with scummvm team :
https://discord.com/channels/581224060529148060/581224061091446795/1030033071120060416

> 1 - I add Sword1 from the scummvm launcher and I modify some keymaps
> 2 - I run the game from this command line : scummvm.exe sword1 , it runs thanks to target, my keymap is OK and everything is fine
> 3 -  I run the game from this command line : scummvm.exe -p c:\MyoldGames\SWORD sword1:sword1 , it runs thanks to the game ID, my keymap is OK but after that the configuration of the game will be deleted in  scummvm.ini  (so the game will not appear in the launcher anymore) 😵‍💫 
>

**Comment by schmurtzm** [comment]
So yes, the scummvm.ini is the way to go ;) 
Then when scummvm.ini is filled, launching the game with `scummvm <target>`  is normally the right way but `scummvm -p <GamePath><target>` seems OK too (just less flexible because the .scummvm file must be in the same directory than the game files)

**Comment by schmurtzm** [comment]
@ner00 I think that we could close this one as `--auto-detect` seems to have bad impact on the scummvm.ini due to a bug in the upstream.  There's still a strange difference of behavior with this option between the core and the standalone but anyway 🤷‍♂️...

As scummvm.ini is the right way to go, I'll create another issue to be able to use the scan option to fill the scummvm.ini automatically. [Something like that](https://github.com/schmurtzm/libretro-scummvm-miyoo-backend/blob/main/libretro.cpp#L395) but implemented in the right way 😋

**Comment by i30817** [comment]
Unfortunately, there is no 'right way' to fill the file automatically because the mass add of upstream is flawed, as i mentioned. It will add bogus entries, which have to be manually deleted. I added bugs in the upstream scummvm tracker about those cases i noticed and where the problem wasn't 'adds languages i don't care about', but the bugs still remain.

A command line flag to only add 'english' or similar would be nice, if you convince upstream to add a '--mass-add' flag to the command line, request something like that too. Something like `--mass-add --lang 'en' .` would add all games in the current directory and subfolders and only add the english games.

**Comment by ner00** [comment]
I do use playlists, so I have no dog in this fight. Besides, I curate my playlists, so... yeah, don't really care about how standalone behaves.

**Comment by schmurtzm** [comment]
Yes you're right but I perfer to scan all the game and remove the one which are not right than doing all manually. The results are not so bad when you do it like that. After that it's just an option of the core, those who want to use it can (as the horrible `--auto-detect` which is already implemented 😆 )

**Comment by schmurtzm** [comment]
> A command line flag to only add 'english' or similar would be nice, if you convince upstream to add a '--mass-add' flag to the command line, request something like that too. Something like `--mass-add --lang 'en' .` would add all games in the current directory and subfolders and only add the english games.

I will ask for that, seems to be a good idea

**Comment by i30817** [comment]
Another problem besides the bogus entries is that there is also no way to _not_ add games where it recognizes the engine but can't recognize the game.

For instance, if you have a unrecognized zcode or ags game and the scummvm scanner recognizes that (it always does) it will add it as 'zcode-#' or 'ags-fallback-#'. In the case of mass add often you don't want that so another switch for '--no-fallback' would be nice too.

This is because those unrecognized games, might or might not work, but their entries have 'names' like 'zcode game' - generic names in short. So you end up having to edit the scummvm.ini or playlists to have sane names for those. If you don't want to bother, a switch for that when 'mass-add' from the command line would also be nice.

**Comment by schmurtzm** [comment]
Wow... and this won't happen if you add the games one by one from the ScummVM launcher ?

**Comment by i30817** [comment]
No it does. What doesn't happen is (almost all) bogus game entries, like ultima8-1 which points to the USECODE subdirectory of the actual game (it doesn't actually run).

If the scanner recognizes the engine but not the game, the game is still added by mass add and will even run (probably). It just has a useless name which you have to correct.

**Comment by schmurtzm** [comment]
OK understood.

`-add --recursive`  and the mass add from the ScummVM launcher make the same isn't it ?

I will ask to the scummvm team (on discord as they don't have github issues) if they can improve `-add --recursive`  feature with a filter by language or avoid to add the game when only the engine is detected.

By the way we will have thousand of people who will use this core when we will release the next version of [Onion](https://github.com/OnionUI/Onion). May be we will have some feedback 😋 

So I still think that `-add --recursive` will be useful even with the current import issues, (as I use a compiled standalone just to do `-add --recursive`  it's a bit of a shame) . I'll create the feature request, do you mind ?

**Comment by i30817** [comment]
No, it's fine.

**Comment by i30817** [comment]
About the language, 'filter by one or more' languages would be better, a way to specify multiple languages, either by repeating the switch or using a separator '--lang en --lang us --lang gb' to cover all the bases of english (scummvm engines don't have much consistency about this and games may be released in 'american english' and 'british english', that's the reason).
